### DISCUSSION TITLE
anime4k实现过程中的疑问

### DISCUSSION THREAD

**Discussion by Arxhexa** [hooke007/mpv_PlayKit#167] [Q&A 问答] [answered]
我参考了以下对AutoDownscalePre作用的解释，但仍然没有完全理解其触发条件以及为什么会有两次upscale。

https://hooke007.github.io/mpv-lazy/[01]_%E7%AC%AC%E4%B8%89%E6%96%B9%E7%9D%80%E8%89%B2%E5%99%A8%E4%BB%8B%E7%BB%8D.html#anime4k-v32-v4x

https://github.com/bloc97/Anime4K/issues/154#issue

按我现在的理解，对于一个1080p视频输出到1440p分辨率应该如下：
源→clamp→restore→upscale(2160p)→downscale_x2(1440p)→downscale_x4(720p)→upscale(1440p)

在downscale_x2过后画面已经与显示分辨率匹配，此时再进行downscale和upscale的意义何在？如果是输出到2160p分辨率那么在第一次upscale过后已经得到了2160p的画面，之后的流程是否是完全多余的性能开销？

**Comment by hooke007** [comment, marked as answer]
作者的演示是基于显示设备2160p，源720p。一次两倍之后1440p，直接做第二次两倍（2880p）显然是性能浪费的有点多，为了节省开销所以在做第二次两倍前预缩回1080p；

假设链路 Anime4K_Upscale_CNN_x2_S;AutoDownscalePre_**x4** 显示设备1440p，源1080p。不会触发你想象的流程，因为不满足hook的条件
流程 a4k 2x upscale(2160p) → mpv dscale(1440p)

AutoDownscalePre 对大多数人来说是用不到的，特殊资源/老片爱好者除外。如上所述，1080p的源 x2之后已经2160p大于大多数人的显示分辨率了，所以构建你的a4k链路的时候不需要像原作者的方案那么复杂（可能也不在乎这点性能浪费）。

> 我那个文档里的说法有点问题。。但是我现在没法直接改
> 
> 假设链路 Anime4K_Upscale_CNN_x2_S;AutoDownscalePre_**x2** 显示设备1080p 源720p
> 
> 执行流程 720p → Anime4K_Upscale_CNN_x2_S (1440p) → AutoDownscalePre_x2 (1080p)
> 这时候就没mpv dscale什么事了，但是我不建议这么用因为 AutoDownscalePre 只是一个简单的双线性，不如用mpv dscale
> 
> 但是它设计这个着色器的考量是假设你要在一次放大后挂更多高开销的着色器，
> 比如链路 Anime4K_Upscale_CNN_x2_S;AutoDownscalePre_x2;Anime4K_Restore_CNN_L
> 预缩回1080p再计算Anime4K_Restore_CNN_L比在1440p计算完Anime4K_Restore_CNN_L后缩回1080p可能要节约的多

**Reply by Arxhexa** [reply]
感谢解答

**Reply by hooke007** [reply]
我那个文档里的说法有点问题。。但是我现在没法直接改

假设链路 Anime4K_Upscale_CNN_x2_S;AutoDownscalePre_**x2** 显示设备1080p 源720p

执行流程 720p → Anime4K_Upscale_CNN_x2_S (1440p) → AutoDownscalePre_x2 (1080p)
这时候就没mpv dscale什么事了，但是我不建议这么用因为 AutoDownscalePre 只是一个简单的双线性，不如用mpv dscale

但是它设计这个着色器的考量是假设你要在一次放大后挂更多高开销的着色器，
比如链路 Anime4K_Upscale_CNN_x2_S;AutoDownscalePre_x2;Anime4K_Restore_CNN_L
预缩回1080p再计算Anime4K_Restore_CNN_L比在1440p计算完Anime4K_Restore_CNN_L后缩回1080p可能要节约的多

**Reply by Arxhexa** [reply]
好的谢谢，确实之前流程中有个双线性scale，还好奇这东西从哪出来的

**Reply by hooke007** [reply]
> 流程中有个双线性scale

如果你是指统计数据里出现的 `scale=bilinear(rgb)` 的话，这个不是 AutoDownscalePre，是vo=gpu实现上的缺陷，会在不同的着色器后强制执行一次mpv的scale=bilinear https://github.com/mpv-player/mpv/issues/5727

**Reply by Arxhexa** [reply]
感谢解答，尝试gpu-next后确实渲染速度有不小提升，按照 https://github.com/bloc97/Anime4K/issues/144 和 https://github.com/bloc97/Anime4K/issues/151 修改滤镜可以在d3d11下（指context和api）实现d3d11va-copy硬解并且正确加载anime4k滤镜

blend-subtitles没法实现个人没看出来有什么太大区别，不过frame timings里会多一个unknown shader，这个您有过了解吗？看着占用的处理时间还挺长的

**Reply by hooke007** [reply]
不知道 AutoDownscalePre 的副产物

**Reply by Arxhexa** [reply]
倒也不是AutoDownscalePre，您第一次回答之后我就只保留了clamp→restore→upscale，使用gpu-next后按照作者说法去除了clamp，仅有restore→upscale

不过画面上其实看起来没什么太大问题

其实好像clamp也可以运行，不过"don't currently work"到底指的是完全不能运行还是运行了没有效果就不太清楚了

**Reply by hooke007** [reply]
> unknown shader
https://github.com/mpv-player/mpv/issues/9411

**Reply by Arxhexa** [reply]
好的谢谢

**Reply by hooke007** [reply]
--blend-subtitles 只要视频宽高比和显示设备不匹配就能看出明显差异。
<img src="https://user-images.githubusercontent.com/41094733/188035015-a21193af-f9e1-432d-962f-92c87273fa82.jpg" height=120> <img src="https://user-images.githubusercontent.com/41094733/188035017-fa4e6cf7-10d2-44a8-ad09-362d26e142ec.jpg" height=120>

此外还涉及色彩和渲染性能的差距（见注释与官方手册）

**Reply by Arxhexa** [reply]
好的，如果真有渲染问题再回退到gpu，相比之下gpu-next更快的渲染速度对我来说更重要一些。
